server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/builder/OnHeapStarTreeBuilder.java

```diff
@@ -269,16 +268,27 @@ public StarTreeDocument next() {
      * Sorts the star-tree documents from the given dimension id
      *
      * @param starTreeDocuments star-tree documents
-     * @param dimensionId id of the dimension
+     * @param dimensionId       id of the dimension
      */
     private void sortStarTreeDocumentsFromDimensionId(StarTreeDocument[] starTreeDocuments, int dimensionId) {
-        Arrays.sort(starTreeDocuments, (o1, o2) -> {
+        Arrays.sort(starTreeDocuments, (doc1, doc2) -> {
             List<Dimension> dimensionsOrder = starTreeField.getDimensionsOrder();
-            for (int i = dimensionId; i < numDimensions; i++) {
-                if (!Objects.equals(o1.dimensions[i], o2.dimensions[i])) {
-                    Dimension dimension = dimensionsOrder.get(i);
-                    Comparator<Long> comparator = dimension.comparator();
-                    return comparator.compare(o1.dimensions[i], o2.dimensions[i]);
+            int totalDimensions = dimensionsOrder.size();
+
+            int currentIndex = dimensionId;
+            for (int i = dimensionId; i < totalDimensions; i++) {
+                Dimension dimension = dimensionsOrder.get(i);
+                int subDimensionsToProcess = dimension.getNumSubDimensions();
+                while (subDimensionsToProcess > 0) {
+                    Long value1 = doc1.dimensions[currentIndex];
```

Comment:
nit : lets avoid unnecessary variable initializations, lets use the actual values directly